src/Uplink/Auth/Token/Token_Manager.php

```diff
@@ -59,44 +61,116 @@ public function validate( string $token ): bool {
 	/**
 	 * Store the token.
 	 *
+	 * @since TBD Added $slug param.
+	 *
 	 * @param  string  $token
+	 * @param  string  $slug The Product slug to store the token for.
 	 *
 	 * @return bool
 	 */
-	public function store( string $token ): bool {
+	public function store( string $token, string $slug = '' ): bool {
 		if ( ! $token ) {
 			return false;
 		}
 
+		$slug = $this->validate_slug( $slug );
+
+		$current_value = $this->get( $slug );
+
 		// WordPress would otherwise return false if the items match.
-		if ( $token === $this->get() ) {
+		if ( $token === $current_value ) {
 			return true;
 		}
 
-		return update_network_option( get_current_network_id(), $this->option_name, $token );
+		$data = $this->get_all();
+
+		if ( ! is_array( $data ) ) {
+			$data = [];
+		}
+
+		$data[ $slug ] = $token;
+
+		return update_network_option( get_current_network_id(), $this->option_name, $data );
 	}
 
 	/**
 	 * Get the token.
 	 *
+	 * @since TBD Added $slug param. Changed return type to mixed.
+	 *
+	 * @param  string  $slug The Product slug to retrieve the token for.
+	 *
 	 * @return string|null
 	 */
-	public function get(): ?string {
-		return get_network_option( get_current_network_id(), $this->option_name, null );
+	public function get( string $slug = '' ): ?string {
+		$slug = $this->validate_slug( $slug );
+
+		$value = get_network_option( get_current_network_id(), $this->option_name, null );
+
+		if ( ! $value ) {
+			return null;
+		}
+
+		if ( ! $slug ) {
+			return is_string( $value ) ? $value : ( array_values( $value )[0] ?? null );
+		}
+
+		if ( is_string( $value ) ) {
+			// Still using old structure, lets return whatever we found.
+			return $value;
+		}
+
+		return $value[ $slug ] ?? null;
+	}
+
+	/**
+	 * Get all the tokens.
+	 *
+	 * @since TBD
+	 *
+	 * @return null|string|array
+	 */
+	public function get_all() {
```

Comment:
Hey @defunctl ! First of all thank you so much for your detailed review.

Have a question about this. The reason i didn't add this in the existing interface (or any other interface) in the first place, is that this method is acting in a way that serves both installations storing a single token or multiple tokens.

So it could return

`null` when no option stored
`string` when still in the schema of storing a single token only
`array` when storing multiple tokens in the same option.

In the interface i cant declare `mixed` as the return type since 7.X version of php don't support it.

I could still create an interface but without type hinting the return value. Let me know what do you think please.

Reply:
Also i don't think we should update the binding in the container. Rather i would update [Token_Manager](https://github.com/stellarwp/uplink/blob/a5fb1fec1abfa969d0e60f2204b9eba88f8e3ced/src/Uplink/Auth/Token/Token_Manager.php) to implement the new interface instead.

Since Token_Manager is a final class i am thinking there shouldn't be any dangers doing that.